File: Data Structures & Algorithms/perfect-squares/submission-5.java
class Solution {
    private int[] dp;

    public int numSquares(int n) {
        dp = new int[n + 1];
        Arrays.fill(dp, -1);
        return solve(n);
    }

    private int solve(int n) {
        if (n == 0)
            return 0;

        if (dp[n] != -1)
            return dp[n];

        int ans = Integer.MAX_VALUE;
        for (int i = (int)Math.sqrt(n); i > 0; i--) {
            ans = Math.min(ans, 1 + solve(n - i * i));
        }
        return dp[n] = ans;
    }
}

File: Data Structures & Algorithms/decode-ways/submission-5.cpp
class Solution {
public:
    int numDecodings(string s) {
        int n = s.size();
        vector<int> dp(n + 2, 0);
        dp[n] = 1;

        for (int i = n - 1; i >= 0; i--) {
            if (s[i] == '0')
                dp[i] = 0;
            else if (s[i] == '1')
                dp[i] = dp[i + 1] + (i < n - 1 ? dp[i + 2] : 0);
            else if (s[i] == '2')
                dp[i] = dp[i + 1] + (i < n - 1 && s[i + 1] <= '6' ? dp[i + 2] : 0);
            else
                dp[i] = dp[i + 1];
        }

        return dp[0];
    }
};